ls1intum/Artemis
# `Programming exercises`: Execute builds on parallel based on system resources for local continuous integration

# Do not deploy on test server, contains config change
### Checklist
#### General


#### Server

### Motivation and Context


Currently, the parallel execution of build jobs in Local CI is determined by the config variable `artemis.continuous-integration.thread-pool-size` which is set to 1 at the moment. That means, only one build job will run concurrently on a single machine. We want to facilitate the concurrent execution of build jobs based on the system resources.

### Description

This PR sets the size of allocated thread pool for the thread pool executor that manages the build jobs on a node based on the number of available CPUs. We reserve 2 CPUs for the application itself and assign 2 CPUs for each build job thread. For example, on a system with 8 CPUs, we can run 3 build jobs concurrently. The thread pool size is always at least 1.
Furthermore, this PR limits the resources available to the build job containers based on the default docker flags set in `artemis.continuous-integration.build` which wasn't the case for Local CI until now.

### Steps for Testing


**Do not deploy on a testserver, as it contains a config change. Test locally with LocalVC Local CI run configuration**
Depending on the number of CPUs your machine has, multiple threads will be allocated for running the build jobs. On a machine with 8 CPUs, you should be able to run 3 build jobs concurrently.

1. Open 2-3 existing Java Gradle programming exercises in different tabs
2. Rerun template and solution builds in all exercises (in quick succession)
![image](https://github.com/ls1intum/Artemis/assets/38919977/36193afd-c630-4d2a-a88c-bec06dad6d35)
3. Depending on how many build jobs your machine can run concurrently, this number of results should come in approximately at the same time.


### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Code Review
#### Manual Tests
